internal/hammer/loadtest/workers.go

```diff
@@ -231,3 +273,178 @@ func (w *LogWriter) Kill() {
 		w.cancel()
 	}
 }
+
+// NewMMDVerifier creates a MMDVerifier.
+func NewMMDVerifier(tracker *client.LogStateTracker, errChan chan<- error, leafMMDChan chan LeafMMD, mmdDuration time.Duration) *MMDVerifier {
+	return &MMDVerifier{
+		tracker:     tracker,
+		errChan:     errChan,
+		leafMMDChan: leafMMDChan,
+		mmdDuration: mmdDuration,
+	}
+}
+
+// MMDVerifier verifies the signed timestamp against the MMD policy for newly
+// added entries by performing inclusion proof.
+type MMDVerifier struct {
+	tracker     *client.LogStateTracker
+	errChan     chan<- error
+	leafMMDChan chan LeafMMD
+	mmdDuration time.Duration
+	cancel      func()
+}
+
+// Run runs the MMD verifier. This should be called in a goroutine.
+func (v *MMDVerifier) Run(ctx context.Context) {
+	if v.cancel != nil {
+		panic("MMDVerifier was ran multiple times")
+	}
+	ctx, v.cancel = context.WithCancel(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		var leafMMD *LeafMMD
+		for {
+			if leafMMD == nil {
+				mmd, ok := <-v.leafMMDChan
+				if !ok {
+					break
+				}
+				leafMMD = &mmd
+			}
+
+			// Retry if the leaf is not yet integrated into the log.
+			if leafMMD.index >= v.tracker.LatestConsistent.Size {
+				// Verify MMD timestamp.
+				sec := int64(leafMMD.timestamp / 1000)
+				nsec := int64((leafMMD.timestamp % 1000) * 1000000)
+				if time.Unix(sec, nsec).Add(v.mmdDuration).Before(time.Now()) {
+					v.errChan <- fmt.Errorf("leaf index %d MMD violation at %d", leafMMD.index, leafMMD.timestamp)
+				}
+
+				break
+			}
+
+			// A new proof builder is required because v.tracker.ProofBuilder
+			// is not thread safe.
```

Comment:
This comment doesn't make much sense in this context?

The proof builder is constructed with a checkpoint, which is super important. As long as the proof builder was only ever used in this worker, you _could_ keep this proof builder about and only invalidate it when the checkpoint from the tracker needs to be bumped. This would be more efficient and if you had a run of entries in a row that could all be proved to be included in the last used checkpoint, then you could avoid some work. However, the state management would be a little more delicate.

Reply:
Right, I hit into the checkpoint change issue. That's why I create a new proof builder with the last checkpoint here. As the performance is doing alright at the moment, I've added a TODO.